### ISSUE TITLE
GPU memory consumption 

### ISSUE THREAD

**Issue by holyseven** [BVLC/caffe#4479]
Hi,

I have searched many issues here and found that caffe consumes GPU memory by storing all data and diff data in GPU (and maybe the weights) of each layer. Is it correct?

I have tried that when fine-tuning a network, if I set the lr_mult of previous 4 or 5 layers to zero, GPU memory consumption could be reduced just a little. It seems that the weights of these layers are not stored any longer but data and diff data are still there.

I am wondering that if the lr_mult of previous layers are zeros, could all data and diff data of these layers be not stored in GPU for saving memory? I think these data will never be used for anything, no?

Thanks in advance,
Jacques.

**Comment by seanbell** [comment]
If you set the lr_mult of a chain of layers, yes, caffe won't compute the backwards pass for those layers.  As long as the forwards pass doesn't use the diff as a temporary storage space, then yes it should also use less memory.  This is a trick used in the Fast R-CNN paper (and others), for example.

Please ask questions about how to use caffe on the mailing list, thanks!

From https://github.com/BVLC/caffe/blob/master/CONTRIBUTING.md:
> _Please do not post usage, installation, or modeling questions, or other requests for help to Issues._
Use the [caffe-users list](https://groups.google.com/forum/#!forum/caffe-users) instead. This helps developers maintain a clear, uncluttered, and efficient view of the state of Caffe.

**Comment by holyseven** [comment]
I'm sorry to ask this question but I think this is about the implementation of caffe using GPU memory. 

And I am still confused. After setting lr_mult to 0, the usage of memory doesn't decrease as much as expected. Let's say, for a case of 5.5~6G GPU usage:

2.5G for storing "data" (forwarding data);
2.5G for "diff" (backwards);
1.0G or less for learnable parameters;
(just my guess)

And for this case, I set lr_mult of all layers to 0, GPU uses still 4.3~4.5G memory. I don't understand. I think it will decrease to around 3.5G.

Is this case reasonable? There is something else which consumes GPU memory? Or "data" and "diff" are not equal at all?


Another question, using the layer softmaxwithloss, or using two layers, softmax layer and cross entropy loss layer, GPU will consume memory very differently? Twice?
The first layer softmax will store immediate data and diff data for loss layer. But softmaxwithloss doesn't have this issue. Am I right?

Thanks
